crates/sui-types/src/crypto.rs

```diff
@@ -262,77 +274,135 @@ impl std::fmt::Debug for Signature {
-            .map_err(|error| SuiError::InvalidSignature {
-                error: error.to_string(),
-            })
-    }
+    const FLAG: [u8; FLAG_LENGTH];
+    const LENGTH: usize = Self::Sig::LENGTH + Self::PubKey::LENGTH + FLAG_LENGTH;
 
-    pub fn get_verification_inputs(
+    fn get_verification_inputs(
         &self,
-        author: SuiAddress,
-    ) -> Result<(AccountSignature, PublicKeyBytes), SuiError> {
+        author: SuiAddress
+    ) -> SuiResult<(Self::Sig, Self::PubKey)> {
         // Is this signature emitted by the expected author?
-        let public_key_bytes: PublicKeyBytes =
-            PublicKeyBytes::from_bytes(self.public_key_bytes()).expect("byte lengths match");
-        let received_addr: SuiAddress = (&public_key_bytes).into();
+        let bytes = self.public_key_bytes();
+        let pk = Self::PubKey::from_bytes(bytes).map_err(|_| SuiError::InvalidSignature {
+            error: format!("Invalid public key"),
+        })?;
+        let received_addr = SuiAddress::from(&pk);
         if received_addr != author {
             return Err(SuiError::IncorrectSigner {
                 error: format!("Signature get_verification_inputs() failure. Author is {author}, received address is {received_addr}")
             });
         }
 
         // deserialize the signature
-        let signature =
-            <AccountSignature as signature::Signature>::from_bytes(self.signature_bytes())
-                .map_err(|err| SuiError::InvalidSignature {
-                    error: err.to_string(),
-                })?;
+        let signature = Self::Sig::from_bytes(self.signature_bytes()).map_err(|err| {
+            SuiError::InvalidSignature {
+                error: err.to_string(),
+            }
+        })?;
+        Ok((signature, pk))
+    }
 
-        // serialize the message (see BCS serialization for determinism)
-        Ok((signature, public_key_bytes))
+    fn new(kp: &Self::KeyPair, message: &[u8]) -> SuiResult<Self> {
+        let sig = kp.try_sign(message)
+            .map_err(|_| SuiError::InvalidSignature { error: "Failed to sign valid message with keypair".to_string() })?;
+
+        let mut signature_bytes: Vec<u8> = Vec::new();
+        signature_bytes.extend_from_slice(&Self::FLAG);
+        signature_bytes.extend_from_slice(sig.as_ref());
+        signature_bytes.extend_from_slice(kp.public().as_ref());
+        Self::from_bytes(&signature_bytes[..]).map_err(|_| SuiError::InvalidSignature {
+            error: format!("Invalid signature"),
+        })
+    }
+}
+
+#[enum_dispatch(Signature)]
+pub trait SuiSignature: Sized {
+    fn bytes(&self) -> &[u8];
+    fn signature_bytes(&self) -> &[u8];
+    fn public_key_bytes(&self) -> &[u8];
+
+    fn flag_bytes(&self) -> &[u8] {
+        &self.bytes()[..FLAG_LENGTH]
+    }
+
+    fn verify<T>(&self, value: &T, author: SuiAddress) -> SuiResult<()>
+    where
+        T: Signable<Vec<u8>>;
+}
+
+
+impl<S: SuiSignatureInner + Sized> SuiSignature for S {
```

Comment:
Is there any other way of cleanly doing this? Looks like the enum dispatch library limits you to defining:
`fn dog(&self) -> det` where `det` is something not generic.